lightning/api_clnraw.go

```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
+	}
+
+	err := l.connection.Call(ctx, Withdraw, []interface{}{address, sats, feerate, minConf}, &reply)
+	if err != nil {
+		return "", err
+	}
+
+	if len(reply.TxID) == 0 {
+		return "", fmt.Errorf("invalid transaction id")
+	}
+
+	return reply.TxID, nil
+}
+
+// calculateExclusion is used since CLN can just exclude channels, so we tell it which channels we want and method calculates the "inverse" of it
+func (l *ClnRawLightningAPI) calculateExclusion(ctx context.Context, outgoingChanIds []uint64) ([]string, error) {
+	result := make([]string, 0)
+
+	if outgoingChanIds == nil || len(outgoingChanIds) == 0 {
+		return nil, nil
+	}
+
+	var info ClnInfo
+	err := l.connection.Call(ctx, GetInfo, []string{}, &info)
+	if err != nil {
+		return nil, err
+	}
+
+	chans, err := l.getMyChannels(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[uint64]struct{})
+	for _, one := range outgoingChanIds {
+		m[one] = struct{}{}
+	}
+
+	for _, one := range chans {
+		_, ok := m[one.ChannelID]
+		if !ok {
+			// not an outgoing channel, create exclusion
+
+			id := FromLndChanID(one.ChannelID)
+
+			peer := ""
+			if one.Node1Pub == info.PubKey {
+				peer = one.Node2Pub
+			} else if one.Node2Pub == info.PubKey {
+				peer = one.Node1Pub
+			} else {
+				glog.Warningf("Channel %d is not mine?", one.ChannelID)
+				continue
+			}
+
+			// direction (u32): 0 if this channel is traversed from lesser to greater id, otherwise 1
+			direction := 0
+			if info.PubKey > peer {
+				direction = 1
+			}
+
+			result = append(result, fmt.Sprintf("%s/%d", id, direction))
+		}
+	}
+
+	return result, nil
+}
+
+// PayInvoice - API call.
+func (l *ClnRawLightningAPI) PayInvoice(ctx context.Context, paymentRequest string, sats int64, outgoingChanIds []uint64) (*PaymentResp, error) {
+	var (
+		err   error
+		reply ClnPayResp
+	)
+
+	exclusions, err := l.calculateExclusion(ctx, outgoingChanIds)
+	fmt.Printf("%+v\n", exclusions)
+	if err != nil {
+		return nil, err
+	}
+
+	if sats > 0 {
+		err = l.connection.Call(ctx, Pay, []interface{}{paymentRequest, sats * 1000, nil, nil, nil, nil, nil, nil, nil, exclusions}, &reply)
+	} else {
+		err = l.connection.Call(ctx, Pay, []interface{}{paymentRequest, nil, nil, nil, nil, nil, nil, nil, nil, exclusions}, &reply)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	r := &PaymentResp{}
+	if reply.Status == "complete" {
+		r.Preimage = reply.PaymentPreimage
+		r.Hash = reply.PaymentHash
+		r.Status = Success
+	} else if reply.Status == "pending" {
+		r.Hash = reply.PaymentHash
+		r.Preimage = ""
+		r.Status = Pending
+	} else {
+		return nil, fmt.Errorf("payment failed")
+	}
+
+	return r, nil
+}
+
+// GetPaymentStatus - API call.
+func (l *ClnRawLightningAPI) GetPaymentStatus(ctx context.Context, paymentHash string) (*PaymentResp, error) {
+	var reply ClnPaymentEntries
+	err := l.connection.Call(ctx, ListPayments, []interface{}{nil, paymentHash}, &reply)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(reply.Entries) != 1 {
```

Comment:
In what case can we get 2 entries here ? Looks a bit dangerous to say it is invalid

Reply:
We sort of call get with an unique id (except that method name is list). So if we get 0 entries the thing doesn't exist but if we get more than 1 something is really screwed up. Maybe we should split the cases for more verbose errors but in any case if the result does not have exactly one entry the function must fail.